detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ForbiddenAnnotation.kt

```diff
@@ -0,0 +1,128 @@
+            "java.lang.Deprecated" to "it is a java annotation. Use `kotlin.Deprecated` instead.",
+            "java.lang.annotation.Documented" to "it is a java annotation. " +
+                "Use `kotlin.annotation.MustBeDocumented` instead.",
+            "java.lang.annotation.Target" to "it is a java annotation. Use `kotlin.annotation.Target` instead.",
+            "java.lang.annotation.Retention" to "it is a java annotation. Use `kotlin.annotation.Retention` instead.",
+            "java.lang.annotation.Repeatable" to "it is a java annotation. Use `kotlin.annotation.Repeatable` instead.",
+            "java.lang.annotation.Inherited" to "Kotlin does not support @Inherited annotation, " +
+                "see https://youtrack.jetbrains.com/issue/KT-22265",
+        )
+    ) { list ->
+        list.associate { it.value to Forbidden(it.value, it.reason) }
+    }
+
+    override fun visitAnnotationEntry(annotation: KtAnnotationEntry) {
+        super.visitAnnotationEntry(annotation)
+        if (annotations.isEmpty()) {
+            return
+        }
+
+        annotation.typeReference?.fqNameOrNull()?.let {
+            check(annotation, it)
+        }
+    }
+
+    override fun visitExpression(expression: KtExpression) {
+        super.visitExpression(expression)
+
+        if (annotations.isEmpty()) {
+            return
+        }
+        expression.expressionTypeOrNull()?.fqNameOrNull()?.let {
+            check(expression, it)
+        }
+    }
+
+    private fun check(element: KtElement, fqName: FqName) {
+        val forbidden = annotations[fqName.asString()]
+
+        if (forbidden != null) {
+            val message = if (forbidden.reason != null) {
+                "The annotation `${forbidden.name}` has been forbidden: ${forbidden.reason}"
+            } else {
+                "The annotation `${forbidden.name}` has been forbidden in the detekt config."
+            }
+            val location = Location.from(element).let { location ->
+                location.copy(
+                    text = location.text.copy(
+                        end = element.children.firstOrNull()?.endOffset ?: location.text.end
+                    )
+                )
+            }
+            report(CodeSmell(issue, Entity.from(element, location), message))
+        }
+    }
+
+    private data class Forbidden(val name: String, val reason: String?)
+
+    private fun KtTypeReference.fqNameOrNull(): FqName? {
+        return if (bindingContext != BindingContext.EMPTY) {
+            bindingContext[BindingContext.TYPE, this]?.fqNameOrNull()
+        } else {
+            null
+        }
+    }
+
+    private fun KtExpression.expressionTypeOrNull(): KotlinType? {
+        return if (bindingContext != BindingContext.EMPTY) {
+            bindingContext[BindingContext.EXPRESSION_TYPE_INFO, this]?.type
+        } else {
+            null
+        }
+    }
```

Comment:
`@RequiresTypeResolution` guarantees non-empty binding context. @BraisGabin this looks like another one for the rule-authors ruleset.

```suggestion
    private fun KtTypeReference.fqNameOrNull(): FqName? =
        bindingContext[BindingContext.TYPE, this]?.fqNameOrNull()

    private fun KtExpression.expressionTypeOrNull(): KotlinType? =
        bindingContext[BindingContext.EXPRESSION_TYPE_INFO, this]?.type
```

Reply:
leaving it open, curious what @BraisGabin thinks

Reply:
Yes, I think that any check of `bindingContext != BindingContext.EMPTY` or `bindingContext == BindingContext.EMPTY` should be flagged as a code smell. That should be an easy rule to implement.